pkg/scheduler/service/cleaner.go

```diff
@@ -34,14 +45,67 @@ func (c *cleaner) Run(ctx context.Context, transition *ClusterStatusTransition,
 		case <-ticker.C:
 			c.purgeReconciliations(transition, config)
 		case <-ctx.Done():
-			c.logger.Info("Stopping cleaner because parent context got closed")
+			c.logger.Info("[CLEANER] Stopping because parent context got closed")
 			ticker.Stop()
 			return nil
 		}
 	}
 }
 
 func (c *cleaner) purgeReconciliations(transition *ClusterStatusTransition, config *CleanerConfig) {
+
+	c.logger.Info("[CLEANER] Process started")
+
+	if config.KeepLatestEntitiesCount > 0 {
+		c.purgeReconciliationsNew(transition, config)
+	} else {
+		c.purgeReconciliationsOld(transition, config)
+	}
+
+	c.logger.Info("[CLEANER] Process finished")
+}
+
+//Purges reconciliations using rules from: https://github.com/kyma-incubator/reconciler/issues/668
+func (c *cleaner) purgeReconciliationsNew(transition *ClusterStatusTransition, config *CleanerConfig) {
+	now := time.Now()
+
+	latestReconciliations, err := c.getLatestReconciliations(config.keepLatestEntitiesCount(), transition)
+	if err != nil {
+		c.logger.Errorf("Cleaner failed to get last %d reconciliations: %s", config.keepLatestEntitiesCount(), err.Error())
+	}
+
+	if len(latestReconciliations) < config.keepLatestEntitiesCount() {
+		//Nothing to clean up
+		return
+	}
+
+	oldestInRange := findOldestReconciliation(latestReconciliations)
+	oldestInRangeAgeDays := diffDays(oldestInRange.Created, now)
+
+	if oldestInRangeAgeDays > config.keepUnsuccessfulEntitiesDays() {
+		//The set of last 'N' reconciliations (which we must keep) contains an entity that is older than configured 'KeepUnsuccessfulEntitiesDays'
+		//It's enough to drop all records older than the oldest from the set.
+		err = c.dropRecordsOlderThan(oldestInRange.Created, transition)
+		if err != nil {
+			c.logger.Errorf("Cleaner failed to remove all reconciliations older than %s: %s", oldestInRange.Created.String(), err.Error())
+		}
+		return
+	}
+
+	//if we're here, there may exist unsuccessful entities older than 'oldestInRange', but within 'KeepUnsuccessfulEntitiesDays' time range.
+	//We have to preserve these (if exist) and remove everything else.
+	deadline := beginningOfTheDay(now.UTC()).AddDate(0, 0, -1*config.keepUnsuccessfulEntitiesDays())
+	err = c.dropRecordsOlderThan(deadline, transition)
+	if err != nil {
+		c.logger.Errorf("Cleaner failed to remove reconciliations older than %s: %s", deadline.String(), err.Error())
+	}
+	err = c.dropSuccessfulRecordsOlderThan(oldestInRange.Created, transition)
```

Comment:
Should we move this `dropSuccessfulRecordsOlderThan` before `if oldestInRangeAgeDays > config.keepUnsuccessfulEntitiesDays() `? I mean this removal doesn't need to depend on that condition, right? We can run this on each cleaning step.

Reply:
TLDR: No, we should not. We could do that, but that would be a waste of time, because we anyway have to execute more broad deletion here: "Delete all records older than [this one]"

Explanation:
One of requirements for this algorithm is: `Keep at most "X" last records, regardless of their state`, where "X" is configurable positive integer.
The other requirement is 'Keep "unsuccessfull" for at least "Y" days back (from now)`, where "Y" is a configurable positive integer. You can write the condition in a more formal way as: `record.status != success && record.age <= Y`

Now consider a situation, where you have the last "X" records, and the oldest of them is from Z days before, where Z > Y.
It means that, with such a set, you satisfy BOTH conditions already: You have last "X" records, and the oldest of them is from Z days before. If there are any "unsuccessful" ones in the set, they all satisfy condition record.age <= Y (because Z > Y)
In that scenario, all you have to do is to drop all remaining records older than the "oldest" in the set, and you're done.
Exactly this scenario is detected in the code by the condition: `if oldestInRangeAgeDays > config.keepUnsuccessfulEntitiesDays()`
It's explained in the commentary.

The other line, `err = c.dropSuccessfulRecordsOlderThan` covers a different scenario.
Consider previous scenario, but now assume Y > Z.
We have three sets now:
1: The set of "X" records we have to keep. The oldest of these is from Z days before. 
2: The set of unsuccessful records where record.age <=Y (Y > Z). This set may partially overlap set number 1, or be distinct, or be empty
3: The set of successful records older than the oldest in set 1.

We have to keep the union of sets 1 and 2: 1 + 2, but we have to remove set 3. This is exactly what `err = c.dropSuccessfulRecordsOlderThan(oldestInRange.Created, transition)` does.